### PR TITLE
Emote Panel TGUI

### PR DESCRIPTION
**Pull request by larentoun**
## About The Pull Request
Adds Emote Panel TGUI in IC category. It shows all the available emotes for the character on the moment of opening the panel. It also has filters for:
* If the emote is visible;
* If the emote is audible;
* If the emote has sound;
* If the emote requires hands;
* If the emote can have params (such as target);

<details><summary>Images</summary>

Screenshot_1
Screenshot_8
Screenshot_2
Screenshot_3
Screenshot_4
Screenshot_5
Screenshot_6
Screenshot_7

</details>

## Why It's Good For The Game
Easier for newer players to use emotes without the use of "*help" and keybindings. Also shows if the emote is visible/audible/etc, so there is less need to spam each of them to see which have sound.

## Changelog
:cl:
add: Emote Panel TGUI added in IC category.
/:cl:

### CODE REVIEW

**Comment by Jacquerel** [comment]
This is a good idea and a good start but, while I don't know exactly in what way we would want it organised, just dumping every emote in a massive block like that is a big eyesore.

**Comment by larentoun** [comment]
> This is a good idea and a good start but, while I don't know exactly in what way we would want it organised, just dumping every emote in a massive block like that is a big eyesore. They should probably be at least put into some kind of equally spaced table.

Yeah, I'm going to go ask around how to make it easier to read or how to properly make it a table

**Comment by Fikou** [comment]
image
can u make lok like tis

**Comment by larentoun** [comment]
> image can u make lok like tis

Thanks for reminding, I was thinking of an ability to add "favourite", but have some struggles with implementation

File: code/modules/emote_panel/emote_panel.dm [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+	return data
+
+/datum/emote_panel/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+	if(.)
+		return
+	switch(action)
+		if("play_emote")
+			var/emote_path = params["emote_path"]
```
Comment (jlsnow301) [inline code/modules/emote_panel/emote_panel.dm:39]: could use a check here to make sure the path is in emote_list

Comment (larentoun) [reply code/modules/emote_panel/emote_panel.dm:39]: just check with `isnull()` ?

Comment (larentoun) [reply code/modules/emote_panel/emote_panel.dm:39]: or just check if it's in the GLOB? It should be there since the TGUI gets items from `emote_list`

**Comment by larentoun** [comment]
Updated images

File: code/modules/emote_panel/emote_panel.dm (hunk 1 above)
Comment (jlsnow301) [reply code/modules/emote_panel/emote_panel.dm:39]: I'm worried they didn't type a path to god-knows-what, but perhaps my paranoia is unfounded. I will look into this later

**Comment by MTandi** [comment]
@larentoun do not highlight the missing attributes in red, they attract too much attention. You can make them half-transparent, or just hide.

File: code/modules/emote_panel/emote_panel.dm [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+			var/use_params = params["use_params"]
+			var/datum/emote/emote = new emote_path()
```
Comment (Mothblocks) [inline code/modules/emote_panel/emote_panel.dm:41]: This code is an enormous vulnerability and allows a crafty user to spawn any object in the entire game. jlsnow caught this but incorrectly doubted themselves. This needs to be fixed. Your PR has been reverted for this reason

**Pull request merged**